### PR TITLE
[GOBBLIN-1001] Implement TimePartitionGlobFinder

### PR DESCRIPTION
**Pull request by zxcware**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. 
    - https://issues.apache.org/jira/browse/GOBBLIN-1001

### Description
- [x] Here are some details about my PR:
   - A `TimePartitionGlobFinder` has the capability to create empty file system dataset if a partition has empty data

### Tests
- [x] My PR adds the following unit tests:
  - `TimePartitionGlobFinder.testDayPartition` covers an empty day partition dataset is created when `TimePartitionGlobFinder.enableEmptyPartition` is true.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by zxcware** [comment]
@autumnust It does use, compositely use instead of inheriting, an existing GlobFinder, which is `DefaultFileSystemGlobFinder`.

**Comment by autumnust** [comment]
> @autumnust It does use, compositely use instead of inheriting, an existing GlobFinder, which is `DefaultFileSystemGlobFinder`.

OK Gotcha. If `enableEmptyPartition` is not enabled, it does nothing more than just setting a different pattern using `DefaultFileSystemGlobFinder`. If it does, it somehow creates empty dataset for yesterday's partitions. This doesn't seem to be a generic finder to me but serving a very specific purpose. Would you consider doing it as a separated extension for GlobFinder internally instead of open-source?

**Comment by zxcware** [comment]
@autumnust Yeah, `yesterdayPartition` is really specific, I'm thinking about generalize it to `enforcePreviousN`(looking for better name suggestions) partitions. Its main responsibility is to create `EmptyFileSystemDataset` if any of the previous N doesn't exist, signaling quiet time. In addition, it focuses on time partitions and supports different time formats(not limitted to `yyyy/MM/dd`) compared to vanilla `DefaultFileSystemGlobFinder`.  (I'm adding comments about it s usage)

By `enforcePreviousN`, it's tied with company requirements even less and makes it more justifiable to open-source. In our use case, we capture the quiet time signal to publish compaction watermark. It can be captured by others to do different operations. 

Another consideration was we have to make internal copies of open source compaction constructs(`MRTask`, `Verifier`, `CompactionAction`), if `EmptyFileSystemDataset` is made internal. Compared to make `EmptyFileSystemDataset` first citizen of open source compaction flow, the implementation and mountainous cost of internalization is high, given most of our pipelines use open source compaction constructs

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/dataset/TimePartitionGlobFinder.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    // Find datasets
+    List<FileSystemDataset> datasets = findDatasets(datasetPattern);
+    // Identify yesterday dataset partitions
+    Set<Path> yesterdayDatasetPartitions = new HashSet<>();
+    datasets.forEach(dataset -> yesterdayDatasetPartitions.add(createYesterdayDatasetPartition(dataset)));
+
+    // Find all dataset time partitions
+    List<FileSystemDataset> datasetPartitions = findDatasets(datasetPartitionPattern);
```
Comment (autumnust) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/dataset/TimePartitionGlobFinder.java]: Could this code block be optimized to be a single expression like : 
`datasetPartitions.addAll(yesterdaypartitions.stream().allMatch(x->!datasetPartitions.contains(x)).map(x -> createEmptyFileDataset))` ?

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/dataset/DefaultFileSystemGlobFinder.java [hunk 2]
```diff
@@ -33,16 +33,6 @@ public DefaultFileSystemGlobFinder(FileSystem fs, Properties properties) throws
   }
 
   public FileSystemDataset datasetAtPath(final Path path) throws IOException {
-    return new FileSystemDataset() {
```
[the remark this answers is not in the document]
Comment (zxcware) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/dataset/DefaultFileSystemGlobFinder.java]: Fixed `PinotAuditCountVerifierTest`. Its usage in `CompactionSuiteBase` will be replaced by a new ser/de mechanism.

**Comment by zxcware** [comment]
@autumnust I generalized the finder which supports look back and will create a virtual partition dataset for any partition within the look back window that doesn't have a physical folder

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/dataset/TimePartitionGlobFinder.java [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+  private final Properties props;
+  private final FileSystem fs;
+
+  public TimePartitionGlobFinder(FileSystem fs, Properties properties) {
+    this(fs, properties,
+        ZonedDateTime.now(ZoneId.of(properties.getProperty(TIME_ZONE, DEFAULT_TIME_ZONE))));
+  }
+
+  @VisibleForTesting
```
[the remark this answers is not in the document]
Comment (zxcware) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/dataset/TimePartitionGlobFinder.java]: yeah, I learned it and meant to use it for documentation purpose if we have found an alternative way

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/dataset/TimePartitionGlobFinder.java [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+    datasets.forEach(dataset -> computedPartitions.addAll(computePartitions(dataset)));
+
+    // This is the final result
+    List<FileSystemDataset> resultPartitions = new ArrayList<>(computedPartitions.size());
+
+    // Find all physical dataset time partitions
+    List<FileSystemDataset> actualPartitions = findDatasets(datasetPartitionPattern);
+
+    String pathStr;
+    for (FileSystemDataset physicalPartition : actualPartitions) {
```
[the remark this answers is not in the document]
Comment (zxcware) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/dataset/TimePartitionGlobFinder.java]: Right. The loop also results in the diff between `computedPartitions` and `actualPartitions`

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/dataset/SimpleFileSystemDataset.java [hunk 5]
```diff
@@ -46,6 +50,9 @@ public String datasetURN() {
     return path.toString();
   }
 
+  /**
+   * @return true if the dataset doesn't have a physical file/folder
+   */
   public boolean getIsVirtual() {
```
Comment (autumnust) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/dataset/SimpleFileSystemDataset.java]: Does it sound more intuitive if `isVirtual` ?

**Pull request closed without merging**